src/core/meshcop/border_agent.cpp

```diff
@@ -225,10 +225,53 @@ BorderAgent::BorderAgent(Instance &aInstance)
     , mTimer(aInstance)
     , mState(kStateStopped)
     , mUdpProxyPort(0)
+#if OPENTHREAD_CONFIG_BORDER_AGENT_ID_ENABLE
+    , mIdInitialized(false)
+#endif
 {
     mCommissionerAloc.InitAsThreadOriginRealmLocalScope();
 }
 
+#if OPENTHREAD_CONFIG_BORDER_AGENT_ID_ENABLE
+Error BorderAgent::GetId(uint8_t *aId, uint16_t &aLength)
+{
+    Error                   error = kErrorNone;
+    Settings::BorderAgentId borderAgentId;
+
+    static_assert(sizeof(mId) == sizeof(borderAgentId), "Invalid Border Agent ID size");
+
+    VerifyOrExit(aLength >= sizeof(mId), error = kErrorInvalidArgs);
+
+    if (mIdInitialized)
+    {
+        memcpy(aId, mId, sizeof(mId));
+        aLength = static_cast<uint16_t>(sizeof(mId));
+        ExitNow();
+    }
+
+    error = Get<Settings>().Read(borderAgentId);
+    if (error == kErrorNone)
+    {
+        mIdInitialized = true;
+        memcpy(mId, borderAgentId.GetId(), sizeof(borderAgentId));
+        memcpy(aId, borderAgentId.GetId(), sizeof(borderAgentId));
+        aLength = static_cast<uint16_t>(sizeof(borderAgentId));
+    }
+    else if (error == kErrorNotFound)
+    {
+        Random::NonCrypto::FillBuffer(mId, sizeof(mId));
+        borderAgentId.SetId(mId, sizeof(mId));
+        SuccessOrExit(error = Get<Settings>().Save(borderAgentId));
+        mIdInitialized = true;
+        memcpy(aId, mId, sizeof(mId));
+        aLength = static_cast<uint16_t>(sizeof(mId));
```

Comment:
Minor nit, but we can check for `kErrorNotFound` first and then have common code for setting up the output values.

Reply:
Updated in new push. Also updated to use `Settings::BorderAgentId mId` directly instead of using a local `Settings::BorderAgentId` id.